src/nvim/state.c

```diff
@@ -66,9 +66,11 @@ void state_enter(VimState *s)
       // Event was made available after the last multiqueue_process_events call
       key = K_EVENT;
     } else {
-      // Duplicate display updating logic in vgetorpeek()
-      if (((State & MODE_INSERT) != 0 || p_lz) && (State & MODE_CMDLINE) == 0
-          && must_redraw != 0 && !need_wait_return) {
+      // Make sure that the screen is fully updated before blocking for new input.
+      // Due to the duality of redraw_later, this can't be done in command line or when we are
+      // waiting for a return key press. In many of those cases the redraw is actually supposed to
+      // happen after the key press, while normally it's used to update the screen immeditely.
```

Comment:
typo, also can be tightened a bit.

```suggestion
      // Ensure the screen is fully updated before blocking for input. Because of the duality of
      // redraw_later, this can't be done in command-line or when waiting for "Press ENTER".
      // In many of those cases the redraw is actually supposed to happen after the key press,
      // while normally it should update the screen immediately.
```